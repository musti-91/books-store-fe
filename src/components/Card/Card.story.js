// Importing global styling
import "../../styles/global.scss";

// JS import
import React from "react";
import { storiesOf } from "@storybook/react";
// Action addons
// import { action } from "@storybook/addon-actions";
// Import of the component
import Card from "./index";

/**
 * Props
 * NOTE: that if you want knobs like [boolean, or text], better to use down directly
 * Otherwise Storybook knobs will not effect changes
 */
const props = {
    book: {
        id: "A1",
        name: "The light of Darkness",
        published: "2019-10-10",
        coverPicture: "https://bit.ly/2NL7gjn",
        rating: 0.1,
        pages: 10,
        awards: [`award`],
        languages: ["English", "French"],
        genre: {
            id: "AE",
            name: "Genre",
        },
        author: {
            id: "AZ",
            name: "My Author",
        },
    },
};
/**
 * Stories: Card
 */
// eslint-disable-next-line no-undef
storiesOf("General.Card", module)
    .addDecorator(storyFn => <div style={{ margin: "2em" }}>{storyFn()}</div>)
    .add("Default", () => <Card {...props} />);
