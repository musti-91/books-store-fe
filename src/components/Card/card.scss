.card {
    width: 22.5em;
    height: 40em;
    border-radius: 0.3em;
    background-color: #fff;
    -webkit-box-shadow: 0px 0px 11px 0px rgba(0, 0, 0, 0.75);
    -moz-box-shadow: 0px 0px 11px 0px rgba(0, 0, 0, 0.75);
    box-shadow: 0px 0px 11px 0px rgba(0, 0, 0, 0.75);
    position: relative;
    transition: all 400ms;
    .cover {
        width: 100%;
        height: 100%;
        border-radius: 0.3em;
        img {
            width: 100%;
            height: 100%;
            border-radius: 0.3em;
        }
    }
    .menu {
        position: absolute;
        width: 2em;
        height: 3em;
        text-align: center;
        line-height: 2.5em;
        font-size: 2.5em;
        top: 50%;
        right: 50%;
        margin-right: -5.2em;
        margin-top: -9em;
        color: #fff;
        background-color: transparent;
        transition: all 300ms;
        transform: rotate(90deg);
    }

    .detail {
        display: flex;
        justify-content: space-evenly;
        flex-wrap: wrap;
        width: 100%;
        height: 20%;
        background-color: #fff;
        border-radius: 0.3em;
        opacity: 0.5;
        transition: all 700ms;
        margin-top: 0.2em;
        .title,
        .genre,
        .info,
        .pages,
        .languages,
        .awards {
            font-family: "Quicksand-Regular";
            font-size: 0.9em;
            background-color: lightgray;
            transition: all 400ms;
            border-radius: 0.3em;
            padding: 0.3em;
            width: 45%;
            margin-bottom: 0.3em;
        }
        .info {
            background-color: red;
            width: 45%;
        }
        .genre {
            background-color: purple;
            width: 45%;
        }
        .pages {
            background-color: yellow;
        }
    }
}
.error {
    color: red;
}
