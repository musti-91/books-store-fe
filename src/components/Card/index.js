// @flow
import React, { type Node } from "react";
import { animated, useSpring } from "react-spring";
// $FlowFixMe: scss files
import "./card.scss";
// mock photo
import { type Book } from "../../data.d";
import photo from "../../styles/images/island_of_fog.jpg";
/**
 * @author
 * @function Card
 **/

type Props = {
    book: Book,
    children?: Node,
};
const Card = ({ children, book }: Props) => {
    const [isShown, setShown] = React.useState(false);
    const styles = useSpring({
        height: isShown ? "75%" : "100%",
        from: { height: "100%" },
    });

    React.useEffect(() => {
        setShown(true);
    }, []);
    return (
        <div className={"card"} onClick={() => setShown(!isShown)}>
            <animated.div style={styles} className="cover">
                <img
                    src={book && book.coverPicture ? book.coverPicture : photo}
                    alt={book && `${book.name}-cover`}
                />
            </animated.div>
            {/* change to icon */}
            <div className={"menu"} onClick={() => setShown(!isShown)}>
                ...
            </div>
            {/* Details for book */}
            {isShown && book && (
                <div>
                    <div className={"detail"}>
                        <div className={"title"}>[]: {book.published}</div>
                        <div className={"info"}>[]:{book.rating}</div>
                        <div className={"pages"}>[]: {book.pages}</div>
                        <div className={"languages"}>
                            []:
                            {book.languages.map(ln =>
                                ln.substring(0, 2).toUpperCase(),
                            )}
                        </div>
                        <div className={"awards"}>
                            []:
                            {book.awards.map(ar => ar)}
                        </div>
                        <div className="genre">
                            []: {book.genre && book.genre.name}
                        </div>
                        <div className={"awards"}>[]: {"Nobel"}</div>
                        <div className="genre">[]: {"Something"}</div>
                    </div>
                </div>
            )}
            {/* eslint-disable react/no-unescaped-entities */}
            {isShown && !book && (
                <div className="error">Can't get Book Detail</div>
            )}
            {children}
        </div>
    );
};

export default Card;
