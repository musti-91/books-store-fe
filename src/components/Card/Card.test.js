/**
 * @jest-environment jsdom
 */
/* eslint-disable react/display-name */
import React from "react";
import { shallow } from "enzyme";
import MockDate from "mockdate";
/**
 *import  Component
 */
import Card from "./index";
/**
 * props
 */
const props = {
    book: {
        id: "A1",
        name: "The light of Darkness",
        published: "2019-10-10",
        coverPicture: "https://bit.ly/2NL7gjn",
        rating: 0.1,
        pages: 10,
        awards: [`award`],
        languages: ["English", "French"],
        genre: {
            name: "my genre name",
        },
        author: {
            name: "My Author",
        },
    },
};
/**
 * Tests: Card Component
 */
describe("Card Component", () => {
    beforeEach(() => {
        MockDate.set("2019-01-01 8:00");
    });
    /**
     * Snapshot:[default]
     */
    test("should match default snapshot", () => {
        expect.assertions(1);

        const wrapper = shallow(<Card {...props} />);

        expect(wrapper).toMatchSnapshot();
    });
});
