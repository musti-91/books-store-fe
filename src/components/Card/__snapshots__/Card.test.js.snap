// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Component should match default snapshot 1`] = `
<div
  className="card"
  onClick={[Function]}
>
  <ForwardRef
    className="cover"
    style={
      Object {
        "height": AnimatedInterpolation {
          "attach": [Function],
          "calc": [Function],
          "children": Array [],
          "detach": [Function],
          "payload": Array [
            AnimatedValue {
              "animatedStyles": Set {},
              "children": Array [],
              "done": true,
              "lastPosition": 0,
              "lastTime": undefined,
              "lastVelocity": undefined,
              "payload": undefined,
              "setValue": [Function],
              "startPosition": 0,
              "startTime": 1546326000000,
              "value": 1,
            },
          ],
        },
      }
    }
  >
    <img
      alt="The light of Darkness-cover"
      src="https://bit.ly/2NL7gjn"
    />
  </ForwardRef>
  <div
    className="menu"
    onClick={[Function]}
  >
    ...
  </div>
</div>
`;
